doc/whats_new/v0.22.rst

```diff
@@ -14,17 +14,56 @@ refer to
 :ref:`sphx_glr_auto_examples_release_highlights_plot_release_highlights_0_22_0.py`.
 
 
+Clear definition of the public API
+----------------------------------
+
+Scikit-learn has a public API, and a private API.
+
+We do our best not to break the public API, and to only introduce
+backward-compatible changes that do not require any user action. However, in
+cases where that's not possible, any change to the public API is subject to
+a deprecation cycle of two minor versions. The private API isn't publicly
+documented and isn't subject to any deprecation cycle, so users should not
+rely on its stability.
+
+Whether a tool is private or public depends on whether you can import it
+without a leading underscore in the import path. For example
```

Comment:
While these refactoring were necessary, I'm happy they happened (and thanks for doing the work), I still prefer the definition,
"""
A function or object is public if it is documented in the reference API and can be imported with an import path without leading underscores.
"""

Just because one can import `from sklearn.utils import issparse` doesn't say anything about `issparse` being part of public API or not. Realistically it's just not possible (and not worth the effort IMO) to remove all such objects from public namespaces. `from sklearn.tests.test_base import MyEstimator` is another example of an object that's definitely not part of the private API.